### DISCUSSION TITLE
CityHub in bad state

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#651] [Q&A] [answered]
My CityHub seems to be in a bad state. Earlier I wanted to flash pybricks on it using Pybricks Code, together with a custom program and a custom name.
As I am used to, I pressed and held the hub button to get into bootloader mode and held it while I tried to flash it. Usually the light turns off and then on again and then I release the button.

However, this time the light did not turn on again and I got an error in pybricks code.:
```
The hub took too long to respond. Restart the hub and try again.
````
After that, I could not press any buttons to get a LED reaction (long press also doesn't seem to turn it off).

After reinserting the batteries the hub turns on when pressing the button and did go into boot loader mode when long pressing ("colorful blinking"). I tried a few times in pybricks code as well as the PoweredUp app to flash some firmware again (also using `make -C bricks/cityhub -j24 deploy`, see the output below) But always when usually the hub should turn on again with colorful blinking it will stay off and stay in this unresponsive state).

Is there a way to somehow "factory reset" the hub? Otherwise It seems like the hub is unrecoverable.

The output from when I tried to update the firmware in powershell (when I managed to get into "colorful" state):
```
PS E:\repos\pybricks-micropython> make -C bricks/cityhub -j24 deploy
make: Entering directory '/cygdrive/e/repos/pybricks-micropython/bricks/cityhub'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
GEN build/genhdr/qstrdefs.collected.h
GEN build/genhdr/compressed.collected
QSTR not updated
Compressed data not updated
Creating firmware...
Searching for CITY hub...
Found: 90:84:2B:E2:76:F2: LEGO System A/S (b'\x00\x00\x00\x10A\x02')
Connecting to 90:84:2B:E2:76:F2: LEGO System A/S (b'\x00\x00\x00\x10A\x02')
Connected successfully!
Erasing flash and starting update
Traceback (most recent call last):
  File "c:\users\lolli\miniconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\users\lolli\miniconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\lolli\miniconda3\Scripts\pybricksdev.exe\__main__.py", line 7, in <module>
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\cli\__init__.py", line 374, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "c:\users\lolli\miniconda3\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "c:\users\lolli\miniconda3\lib\asyncio\base_events.py", line 642, in run_until_complete
    return future.result()
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\cli\flash.py", line 323, in flash_firmware
    await flash_ble(hub_kind, firmware, metadata)
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\cli\flash.py", line 273, in flash_ble
    await updater.flash(firmware, metadata)
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\flash.py", line 311, in flash
    response = await self.bootloader_request(
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\flash.py", line 257, in bootloader_request
    await self.write(data, request.write_with_response)
  File "c:\users\lolli\miniconda3\lib\site-packages\pybricksdev\ble\__init__.py", line 143, in write
    await self.client.write_gatt_char(
  File "c:\users\lolli\miniconda3\lib\site-packages\bleak\backends\winrt\client.py", line 611, in write_gatt_char
    _ensure_success(
  File "c:\users\lolli\miniconda3\lib\site-packages\bleak\backends\winrt\client.py", line 104, in _ensure_success
    raise BleakError(f"{fail_msg}: Unreachable")
bleak.exc.BleakError: Could not write value bytearray(b'D\x18\xa9\x02\x00') to characteristic 000D: Unreachable
make: *** [../stm32/stm32.mk:686: deploy] Error 1
make: Leaving directory '/cygdrive/e/repos/pybricks-micropython/bricks/cityhub'
```

Interestingly, now that I think about it I don't think I could successfully flash the firmware using pybricks code since at least some weeks. I always had to resort to the command line to flash pybricks. In pybricks code, after connecting to the bootlader, the LED usually does turn off and on again, but pybricks code reports an error and does not flash the firmware. This usually doesn't leave the hub in a bad state. I only used pybricks code in MS Edge these past few weeks.

I just tried again in Google Chrome using a different cityhub (I can flash it in the command line), and I still have the same problem:

```
The hub was disconnected before flashing was completed. Restart the hub and try again.
```
This should probably be a sperate issue.

**Comment by laurensvalk** [comment, marked as answer]
Is there a sensor or motor plugged into port A by any chance?

There’s a bug in the bootloader that prevents the update from starting if a UART device is plugged in.

When you hit this, the light turns off but the hub is still on. So, remove the sensor, remove the batteries, and then try again.

**Reply by Novakasa** [reply]
It could be that there was a motor plugged in when I tried the first time, but since then I tried many times without anything plugged in with no success.

**Reply by dlech** [reply]
Also make sure the the batteries are not low.

**Reply by dlech** [reply]
Or try a different computer or Android device.

**Reply by Novakasa** [reply]
I just tried again after a PC restart and now it works (via command line). Did not expect that because other hubs worked fine even before the restart, but it does seem like the PC was in a bad state here.

Thanks for the suggestions!

I'll probably open an issue soon about firmware flashing not working in pybricks code for me.

**Reply by Novakasa** [reply]
One thing that seems weird to me however is that the PoweredUp app also did not work... Maybe it works now because I had the hub without power for a longer time.
